devex/vscode-extension/assets/edgeSolutionTemplateFolder/modules/[[project-name]]/host/main.c

```diff
@@ -0,0 +1,235 @@
+#include <stdlib.h>
+
+#include <openenclave/host.h>
+
+#include "iothub_module_client_ll.h"
+#include "iothub_client_options.h"
+#include "iothub_message.h"
+#include "azure_c_shared_utility/threadapi.h"
+#include "azure_c_shared_utility/crt_abstractions.h"
+#include "azure_c_shared_utility/platform.h"
+#include "azure_c_shared_utility/shared_util_options.h"
+#include "iothubtransportmqtt.h"
+#include "iothub.h"
+#include "time.h"
+
+int create_enclave(int argc, const char* argv[]);
+int terminate_enclave();
+int call_enclave(char *input_msg, char *enclave_msg, unsigned int enclave_msg_size);
+
+typedef struct MESSAGE_INSTANCE_TAG
+{
+    IOTHUB_MESSAGE_HANDLE messageHandle;
+    size_t messageTrackingId;  // For tracking the messages within the user callback.
+}
+MESSAGE_INSTANCE;
+
+size_t messagesReceivedByInput1Queue = 0;
+
+// SendConfirmationCallback is invoked when the message that was forwarded on from 'InputQueue1Callback'
+// pipeline function is confirmed.
+static void SendConfirmationCallback(IOTHUB_CLIENT_CONFIRMATION_RESULT result, void* userContextCallback)
+{
+    // The context corresponds to which message# we were at when we sent.
+    MESSAGE_INSTANCE* messageInstance = (MESSAGE_INSTANCE*)userContextCallback;
+    printf("Confirmation[%zu] received for message with result = %d\r\n", messageInstance->messageTrackingId, result);
+    IoTHubMessage_Destroy(messageInstance->messageHandle);
+    free(messageInstance);
+}
+
+// Allocates a context for callback and clones the message
+// NOTE: The message MUST be cloned at this stage.  InputQueue1Callback's caller always frees the message
+// so we need to pass down a new copy.
+static MESSAGE_INSTANCE* CreateMessageInstance(char* messageBody)
+{
+    MESSAGE_INSTANCE* messageInstance = (MESSAGE_INSTANCE*)malloc(sizeof(MESSAGE_INSTANCE));
+    if (NULL == messageInstance)
+    {
+        printf("Failed allocating 'MESSAGE_INSTANCE' for pipelined message\r\n");
+    }
+    else
+    {
+        memset(messageInstance, 0, sizeof(*messageInstance));
+
+        if ((messageInstance->messageHandle = IoTHubMessage_CreateFromString(messageBody)) == NULL)
+        {
+            free(messageInstance);
+            messageInstance = NULL;
+        }
+        else
+        {
+            IoTHubMessage_SetContentTypeSystemProperty(messageInstance->messageHandle, "application%2fjson");
+            IoTHubMessage_SetContentEncodingSystemProperty(messageInstance->messageHandle, "utf-8");
+
+            messageInstance->messageTrackingId = messagesReceivedByInput1Queue;
+        }
+    }
+
+    return messageInstance;
+}
+
+static IOTHUBMESSAGE_DISPOSITION_RESULT SendEnclaveResponse(IOTHUB_MODULE_CLIENT_LL_HANDLE iotHubModuleClientHandle, char* messageBodyStr)
+{
+    IOTHUBMESSAGE_DISPOSITION_RESULT result;
+    IOTHUB_CLIENT_RESULT clientResult;
+
+    char* enclaveMessage = (char*)malloc(512 * sizeof(char));
+    int enclaveResult = call_enclave(messageBodyStr, enclaveMessage, 512);
```

Comment:
should this check for null enclaveMessage before doing call_enclave?

Reply:
Fixed.